### PR TITLE
Feature/HIR-22. Endpoints to manage availability restrictions

### PR DESCRIPTION
**Pull request by oleg1std**
Add endpoints to ADD, GET, DELETE room availability restrictions.
ADD, GET, DELETE  api/1.0/management/contracts/contracts/{contractId}/availability-restrictions

### CODE REVIEW

File: HappyTravel.Hiroshima.Data/Models/Rooms/RoomAvailabilityRestriction.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class RoomAvailabilityRestriction
+    {
+        public int Id { get; set; }
+        public int RoomId { get; set; }
+        public DateTime FromDate { get; set; }
+        public DateTime ToDate { get; set; }
+        public AvailabilityRestrictions Restriction { get; set; }
+        public int ContractId { get; set; }
+    }
```
Comment (hasmikharutyunyan) [inline HappyTravel.Hiroshima.Data/Models/Rooms/RoomAvailabilityRestriction.cs:14]: As I understand here we use foreign keys, maybe we need to add `Contract` ?

File: HappyTravel.Hiroshima.DirectManager/RequestValidators/AvailabilityRestrictionsValidator.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            RuleFor(restrictions => restrictions)
+                .Must(availabilityRestrictions => AreRestrictionsInAscendingOrder(availabilityRestrictions, contract))
+                .WithMessage(restrictions => "Invalid date ranges of the availability restrictions");
+            RuleForEach(restrictions => restrictions).SetValidator(new AvailabilityRestrictionValidator());
+        }
+
+
+        private bool AreRestrictionsInAscendingOrder(List<Models.Requests.AvailabilityRestriction> availabilityRestrictions, Contract contract)
+        {
+            var availabilityRestrictionGroups = availabilityRestrictions.GroupBy(availabilityRestriction => availabilityRestriction.RoomId).ToDictionary(group=>group.Key, group => group.OrderBy( restriction => restriction.FromDate).ToList());
```
Comment (hasmikharutyunyan) [inline HappyTravel.Hiroshima.DirectManager/RequestValidators/AvailabilityRestrictionsValidator.cs:21]: needed spaces `group => group.Key `

File: HappyTravel.Hiroshima.DirectManager/Services/AvailabilityRestrictionsManagementService.cs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                if (restriction != null)
+                {
+                    availabilityRestrictions = availabilityRestrictions.Where(availabilityRestriction => availabilityRestriction.Restriction == restriction.Value);
+                }
+
+                return await availabilityRestrictions
+                    .OrderBy(availabilityRestriction => availabilityRestriction.Id)
+                    .Skip(skip)
+                    .Take(top)
+                    .ToListAsync();
```
Comment (hasmikharutyunyan) [inline HappyTravel.Hiroshima.DirectManager/Services/AvailabilityRestrictionsManagementService.cs:107]: As I understand, all above are queries (`availabilityRestrictions`, `validRoomIds`) . Maybe will be understandable to add `query` in variable name

File: HappyTravel.Hiroshima.Data/Models/Contract.cs [hunk 4]
```diff
@@ -10,5 +11,8 @@ public class Contract
         public string Name { get; set; }
         public string Description { get; set; }
         public int ContractManagerId { get; set; }
+        public List<Seasons.Season> Seasons { get; set; } 
+        public List<Rooms.RoomPromotionalOffer> PromotionalOffers { get; set; } 
+        public List<Rooms.RoomAvailabilityRestriction> RoomAvailabilityRestriction { get; set; }
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Hiroshima.Data/Models/Contract.cs:16]: There are relations between the contract and these tables.

File: HappyTravel.Hiroshima.Common/Models/Accommodations/Rooms/AvailabilityRestrictions.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿﻿namespace HappyTravel.Hiroshima.Common.Models.Accommodations.Rooms
+{
+    public enum AvailabilityRestrictions
```
[the remark this answers is not in the document]
Comment (oleg1std) [reply HappyTravel.Hiroshima.Common/Models/Accommodations/Rooms/AvailabilityRestrictions.cs:3]: System.Text.Json is used in this project. There are no attributes like [JsonConverter(typeof(StringEnumConverter))]

**Review by hasmikharutyunyan — APPROVED** [review]

**Pull request merged**